### ISSUE TITLE
[BUG] Mute Group and feedback

### ISSUE THREAD

**Issue by JohnenB** [bitfocus/companion-module-behringer-x32#112]
I  am experiencing an issue with my Behringer X32. The easiest way to explain it is through the attached video. 

If I mute a specific channel, I can see the change both on the X32 software and on my Stream Deck. However, when I use a mute group, I can only see whether the channels are muted or not on the X32 software. The Stream Deck updates with the correct mute state only when I adjust a fader or perform a similar action. 

Additionally, I've observed that if I create a button that toggles the mute on two or more channels, I need to have a minimum delay of 101ms between the actions in Companion. Otherwise, I encounter the same issue. 

I am running Companion on a Windows 10 PC, and my Companion version is 3.1.2. I hope this clarifies my issue. 

https://github.com/bitfocus/companion-module-behringer-x32/assets/150924951/00f572b6-ba43-4219-96f0-880e9b250721

**Comment by CalvarySteve** [comment]
I am experiencing the same thing. My Feedback of MUTE GROUP Status is not working. I cannot get consistent FEEDBACK from my X32 on the status of my Mute Groups

**Comment by psycomics** [comment]
Also having the same issue on 3.1.0 release as @JohnenB - issue not present on companion version 2.4.2.  On X32 Rack on firmware 4.06-8.  Will try and fix as is preventing us from upgrading to companion V3.

When unmuting a single channel feedback works fine; issue appears when doing multiple mutes at <101 ms delay between mutes.  

I can't see a mute variable for a channel listed in the companion variables stored (unlike fader positions) so assume it's pulled every time and is a update time issue? Any other ideas of where to check?

**Comment by ploveman** [comment]
Just a point of difference, the an x32 simulator and x32 module version 3.1.0 the feedback seems to work properly for me when I mute/unmute a mute group.  Next time I'm on our church actual x32 I'll try and see if it behaves any differently (I am just using the GUI/emulator if that sheds light on anything).

**Comment by ploveman** [comment]
> Just a point of difference, the an x32 simulator and x32 module version 3.1.0 the feedback seems to work properly for me when I mute/unmute a mute group. Next time I'm on our church actual x32 I'll try and see if it behaves any differently (I am just using the GUI/emulator if that sheds light on anything).

I checked on our x32 at church today and the feedback for mute group seemed to work as I expected (again on GUI emulator vs streamdeck) with 4.06-8 firmware

**Comment by Gripenhammar** [comment]
Same issue here. I also found the delay workaround, and found out that when muting eight channels with the same button, you can set (with relative delays) a 200 ms on command row #2, and another 200 ms on command row #8, and they will all get feedback. But still a bit of a nuisance.

**Comment by ki3v** [comment]
I have what I believe is the same or a related issue.  I am using a Razer Stream Controller (a rebadged Loupedeck Live), with Companion 3.3.1 and X32/M32 module v3.1.0.  

I have several buttons with mute actions and feedback, and another button that has several actions (mutes one mic, unmutes another, and fires off some non-X32 actions as well.  Most of the time, when I hit the third multi-action button, the mute feedback on the two mic buttons does not update.  However, if I tap either of them, both update to the correct feedback.  

I have one of the dials set to change my main LR volume, and if I adjust it, both buttons' feedback does update, even though they are on unrelated channels.

I'd be happy to do any testing needed to work on this issue.  I don't know much JS/TS but I do some development work and am familiar with build systems, VSCode, etc so I can set up build environments and do some reasonably sophisticated testing if it's helpful.

**Comment by ki3v** [comment]
Also, this seems to be a duplicate with issue #98.

**Comment by ploveman** [comment]
> I have what I believe is the same or a related issue. I am using a Razer Stream Controller (a rebadged Loupedeck Live), with Companion 3.3.1 and X32/M32 module v3.1.0.
> 
> I have several buttons with mute actions and feedback, and another button that has several actions (mutes one mic, unmutes another, and fires off some non-X32 actions as well. Most of the time, when I hit the third multi-action button, the mute feedback on the two mic buttons does not update. However, if I tap either of them, both update to the correct feedback.
> 
> I have one of the dials set to change my main LR volume, and if I adjust it, both buttons' feedback does update, even though they are on unrelated channels.
> 
> I'd be happy to do any testing needed to work on this issue. I don't know much JS/TS but I do some development work and am familiar with build systems, VSCode, etc so I can set up build environments and do some reasonably sophisticated testing if it's helpful.
> 
> Edit: I just noticed that if I use the front panel channel level control of the X32 (Rack model so no faders per se) or the physical mute button, the Companion surface button updates immediately. Even if I make an unrelated change, such as the main LR level knob on the X32 Rack front panel, the mute feedback on the Companion surface updates correctly.
> 
> So I seem to be observing that when a Companion surface button is pressed, it _can_ in some cases cause another button's feedback not to update, but events outside Companion seem to trigger feedback successfully.

From what I can tell from the code; I think the feedback comes from the OSC notification regardless of the source of the change.  I've tried with an x32 emulator and my x32 at church and haven't been able to reproduce it with companion GUI (I don't have a streamdeck or any external hardware).  Like you I'm pretty new to TS, but I've managed to add logging to look at other bits so if there's a consistent way to reproduce I might be able to at least debug it.

**Comment by ki3v** [comment]
When you say X32 emulator, are you referring to the one by Patrick-Gilles Maillot?  I have not heard of an X32 emulator before and that's what comes up for me on a quick search.

I may give the emulator a try, but I do have an X32 rack sitting on my desk that I can test with as needed.

I'm just speculating of course, but I was wondering if OSC replies that come in too quickly get swallowed by the code somehow when you are pressing buttons.  But of course only in certain situations, and now always.  

For me the fairly consistent behavior is that if I press a button other than the one with the mute feedback that causes that mute to change on the X32, the mute feedback usually doesn't fire.

If that's the case, it seems likely that it's not actually directly related to the mute feedback, but any feedback from OSC updates... but I feel like the channel mute feedback is probably by far the most common for buttons.  I guess we could/should also test mute group feedback etc.

**Comment by ploveman** [comment]
> When you say X32 emulator, are you referring to the one by Patrick-Gilles Maillot? I have not heard of an X32 emulator before and that's what comes up for me on a quick search.
> 
> I may give the emulator a try, but I do have an X32 rack sitting on my desk that I can test with as needed.
> 
> I'm just speculating of course, but I was wondering if OSC replies that come in too quickly get swallowed by the code somehow when you are pressing buttons. But of course only in certain situations, and now always.
> 
> For me the fairly consistent behavior is that if I press a button other than the one with the mute feedback that causes that mute to change on the X32, the mute feedback usually doesn't fire.
> 
> If that's the case, it seems likely that it's not actually directly related to the mute feedback, but any feedback from OSC updates... but I feel like the channel mute feedback is probably by far the most common for buttons. I guess we could/should also test mute group feedback etc.

Yes, that's the emulator I've been playing with (kudos to him for that), I know it may not be 1:1 but so far it's been pretty good for some of my dev testing (since my x32 is at church).  I'm not sure what I'm doing differently but I'm seeing the feedback regardless of whether I change on the mixer directly or from another button or in my home testing with the x32 edit app.  If it's some sort of race condition that might explain why I'm getting different results, but also can be a lot harder to sort out.

**Comment by ploveman** [comment]
I don't know 100% if it would be helpful, but if someone with either of the issues could do a packet capture on the PC to the mixer might indicate if we're seeing a difference in messages or a difference in processing in companion.  https://www.wireshark.org/ is what I've used to take a capture.  I usually do the capture across all IP interfaces to make sure it's captured regardless of the NIC and then save as a pcapng.  If possible it'd be nice to have it have a host filter during the capture limited to the mixer (so only capturing those packets)...example host 192.168.100.100 (mixer at our church IP) and try and start the capture then push whatever to reproduce what you'd expect feedback on.  Either way please mention what your PC and mixer IPs are so it's easy to find what's sending what.  
[x32_module_startup.pcapng.gz](https://github.com/user-attachments/files/16325936/x32_module_startup.pcapng.gz)
I'm attaching a capture I took this morning from the enabling of the connection to the mixer (just to see what that looks like) and then a few general mute changes.  This is just for reference if anyone wants to look at the OSC between companion and the mixer.

**Comment by Julusian** [comment]
I've just taken a look and this appears to be a simple fix https://github.com/bitfocus/companion-module-behringer-x32/commit/89aab36f6223c6b8ae78fac478148a6c4282ba2f
I think thats a bug that would have been introduced when updating the module for 3.0, as that had some changes in the module api.

I don't know why it worked for muting a single channel and not a group, but I didn't look into that.
The fix will be in the betas shortly

**Comment by ploveman** [comment]
> I've just taken a look and this appears to be a simple fix [89aab36](https://github.com/bitfocus/companion-module-behringer-x32/commit/89aab36f6223c6b8ae78fac478148a6c4282ba2f) I think thats a bug that would have been introduced when updating the module for 3.0, as that had some changes in the module api.
> 
> I don't know why it worked for muting a single channel and not a group, but I didn't look into that. The fix will be in the betas shortly

Does your fix have an indication of why some aren't experiencing the bug?  Does it possibly relate to what feedbacks are configured?

**Comment by ki3v** [comment]
Sorry to have been quiet on this for so long.  This fix fixed the issue for me.  Thanks!
